controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java

```diff
@@ -563,11 +564,50 @@ public void listStreams(final String scopeName, final String showInternalStreams
             LoggerHelpers.traceLeave(log, "listStreams", traceId);
             return;
         }
-        boolean showOnlyInternalStreams = showInternalStreams != null && showInternalStreams.equals("true");
-        controllerService.listStreamsInScope(scopeName, requestId)
-                .thenApply(streamsList -> {
-                    StreamsList streams = new StreamsList();
-                    streamsList.forEach((stream, config) -> {
+        boolean showOnlyInternalStreams = filterType != null && filterType.equals("showInternalStreams");
+        boolean showStreamsWithTag = filterType != null && filterType.equals("tag");
+        String tag = null;
+        if (showStreamsWithTag && filterValue != null) {
+            tag = filterValue;
+        }
+        if (tag != null) {
+            List<Stream> streams = new ArrayList<>();
+            StreamsList responseStreams = new StreamsList();
+            String finalTag = tag;
+            localController.listStreamsForTag(scopeName, tag).collectRemaining(s -> streams.add(s)).thenApply(s -> streams).thenApply(s -> {
```

Comment:
```suggestion
                  localController.listStreamsForTag(scopeName, tag)
                           .collectRemaining(streams::add)
                           .thenCompose(s -> {
                               List<CompletableFuture<ImmutablePair<Stream, StreamConfiguration>>> futureList = streams.stream().filter(stream -> {
                                   boolean isAuthorized = false;
                                   try {
                                       isAuthorized = restAuthHelper.isAuthorized(authHeader, authorizationResource.ofStreamInScope(scopeName, stream.getStreamName()),
                                                                                  principal, READ);
                                   } catch (AuthException e) {
                                       log.warn(requestId, "List Streams with tag {} for scope {} failed due to authentication failure.",
                                                finalTag, scopeName);
                                       // Ignore. This exception occurs under abnormal circumstances and not to determine
                                       // whether the user is authorized. In case it does occur, we assume that the user
                                       // is unauthorized.
                                   }
                                   return isAuthorized;
                               }).map(stream -> localController.getStreamConfiguration(scopeName, stream.getStreamName())
                                                               .thenApply(cfg -> new ImmutablePair<>(stream, cfg)))
                                 .collect(Collectors.toList());
                               return Futures.allOfWithResults(futureList);
                           }).thenApply(streamCfgPairs -> {
                               StreamsList responseStreams = new StreamsList();
                               streamCfgPairs.forEach(p -> responseStreams.addStreamsItem(ModelHelper.encodeStreamResponse(p.left.getScope(), p.left.getStreamName(), p.right)));
                               log.info(requestId, "Successfully fetched streams for scope: {} with tag: {}", scopeName, finalTag);
                               return Response.status(Status.OK).entity(responseStreams).build();
                           }).exceptionally(exception -> {
```

Reply:
Done